net: lwm2m: Fix address length at connect

biglben opened the pull request:
If NET_IPV4 and NET_SOCKETS_PACKET is enabled, NET_SOCKADDR_MAX_SIZE will be bigger than the ipv4 address length. This is a problem when DTLS is used as the address comparison will fail because of the different length of the received and the stored address.
This is also a problem if NET_IPV6 and NET_IPV4 is enabled and the remote address is a ipv4 address.

Signed-off-by: Benjamin Bigler <benjamin.bigler@securiton.ch>

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -5663,7 +5663,7 @@ int lwm2m_socket_start(struct lwm2m_ctx *client_ctx)
 
 	if (connect(client_ctx->sock_fd, &client_ctx->remote_addr,
-		    NET_SOCKADDR_MAX_SIZE) < 0) {
+	        client_ctx->addr_len) < 0) {
```
Perhaps, instead of storing the address size in the context, just resolve the size here based on `client_ctx->remote_addr.sa_family`?

biglben committed:
net: lwm2m: Fix address length at connect

If NET_IPV4 and NET_SOCKETS_PACKET is enabled, NET_SOCKADDR_MAX_SIZE will
be bigger than the ipv4 address length.
This is a problem when DTLS is used as the address comparison will fail
because of the different length of the received and the stored address.
This is also a problem if NET_IPV6 and NET_IPV4 is enabled and the remote
address is a ipv4 address

Signed-off-by: Benjamin Bigler <benjamin.bigler@securiton.ch>

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -5661,9 +5662,19 @@ int lwm2m_socket_start(struct lwm2m_ctx *client_ctx)
 
 	if (connect(client_ctx->sock_fd, &client_ctx->remote_addr,
-		    NET_SOCKADDR_MAX_SIZE) < 0) {
+		addr_len) < 0) {
```
Please keep the alignment.
```suggestion
		    addr_len) < 0) {
```

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -5661,9 +5662,19 @@ int lwm2m_socket_start(struct lwm2m_ctx *client_ctx)
+	}
+	else if ((client_ctx->remote_addr).sa_family == AF_INET6) {
+		addr_len = sizeof(struct sockaddr_in6);
+	}
+	else {
```
```suggestion
	} else {
```

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -5661,9 +5662,19 @@ int lwm2m_socket_start(struct lwm2m_ctx *client_ctx)
 #endif /* CONFIG_LWM2M_DTLS_SUPPORT */
+	if ((client_ctx->remote_addr).sa_family == AF_INET) {
+		addr_len = sizeof(struct sockaddr_in);
+	}
+	else if ((client_ctx->remote_addr).sa_family == AF_INET6) {
```
```suggestion
	} else if ((client_ctx->remote_addr).sa_family == AF_INET6) {
```

SeppoTakalo reviewed (changes requested):
Thank you for the contribution.
I'll suggest that you change the coding style to match existing.
